crypto/cms/cms_sd.c

```diff
@@ -227,6 +227,30 @@ int ossl_cms_SignerIdentifier_cert_cmp(CMS_SignerIdentifier *sid, X509 *cert)
         return -1;
 }
 
+/* Method to map any, incl. provider-implemented PKEY types to OIDs */
+static int cms_generic_sign(CMS_SignerInfo *si, int verify)
+{
+    if (!verify) {
+        int hnid;
+        X509_ALGOR *alg1, *alg2;
+        EVP_PKEY *pkey = si->pkey;
+        const char *typename = EVP_PKEY_get0_type_name(pkey);
+
+        CMS_SignerInfo_get0_algs(si, NULL, NULL, &alg1, &alg2);
+        if (alg1 == NULL || alg1->algorithm == NULL)
+            return -1;
+        hnid = OBJ_obj2nid(alg1->algorithm);
+        if (hnid == NID_undef)
+            return -1;
+
+        /* Generic PKEY->name->NID->OID mapping check */
+        if (typename == NULL)
+            return -1;
+        return X509_ALGOR_set0(alg2, OBJ_nid2obj(OBJ_sn2nid(typename)), V_ASN1_UNDEF, NULL);
```

Comment:
I suggest replacing `OBJ_nid2obj(OBJ_sn2nid(typename))` with `OBJ_txt2obj(typename, 0)`

Reply:
Done in https://github.com/openssl/openssl/pull/17733/commits/e2f121555580330d4991b9894912fa336f5e8949.